python/tvm/relax/training/optimizer.py

```diff
@@ -0,0 +1,582 @@
+        elif not isinstance(params, list):
+            params = [params]
+        self._state = None
+        self._dtype = None
+        self._check_params_and_dtype(params)
+        self._param_list = params
+        assert len(self._param_list) == len(
+            set(self._param_list)
+        ), "There exist duplicate Vars in the parameter list."
+
+    def _check_params_and_dtype(self, params: List[Var]):
+        for x in params:
+            assert isinstance(x, Var), "Not every parameter is Var."
+            assert isinstance(x.struct_info, TensorStructInfo), "Not every parameter is Tensor Var"
+            data_type = tvm.DataType(x.struct_info.dtype)
+            assert data_type.type_code in (
+                tvm.DataTypeCode.BFLOAT,
+                tvm.DataTypeCode.FLOAT,
+            ), "Pamameters must be of float dtype"
+            if self._dtype is None:
+                self._dtype = x.struct_info.dtype
+            else:
+                assert (
+                    self._dtype == x.struct_info.dtype
+                ), "All parameters should have the same dtype"
+
+    def add_params(self, params: Union[Var, List[Var]]):
+        """Append one parameter or a list of new parameters to the optimizer.
+
+        Parameters
+        ----------
+        params : Union[Var, List[Var]]
+            The parameter or the list of parameters to append.
+
+            Parameters should all be Vars of floating point Tensors, including float32, float64,
+            float16, bfloat16, etc. Currently, all parameters should have the same dtype, and that
+            dtype will be used as the dtype of the optimizer states.
+
+        Note
+        ----
+        This method can only be called before `opt.state` or `opt.get_function()`.
+        """
+        assert self._state is None, "Add parameter after the state is acquired"
+        if not isinstance(params, list):
+            params = [params]
+        self._check_params_and_dtype(params)
+        self._param_list += params
+        assert len(self._param_list) == len(
+            set(self._param_list)
+        ), "There exist duplicate Vars in the parameter list."
+
+    @property
+    def state(self):
+        """Return the state of the optimizer. This should be used as the last argument of the
+        function that `get_function()` returns, and its new value is returned as the last return
+        value of this function.
+
+        The state of an optimizer will be constructed when `opt.state` is called for the first time.
+        Before that, you can freely add parameters using `opt.add_params()`.
+
+        Returns
+        -------
+        res : ADT
+            An ADT object representing the state of the optimizer.
+        """
+        return self._state
+
+    @state.setter
+    def state(self, value):
+        """Setter of state.
+
+        If `state` property is overloaded, `state` setter must be overloaded at the same time.
+        """
+        self._state = value
```

Comment:
I’m a bit perplexed about the design here. Now we have the attribute `_state`, a property `state` that returns and lazily initializes the state, and a method `state(...)` that set the state. Could you explain a bit about the purpose here?

It doesn’t seem very clean to have a same name with multiple occurrences, with each of them having different usage. Intuitively we can have `get_state` and `set_state`.

Reply:
I see. I didn’t know the property setter in Python before. Now it looks good to me.

Shall we rename `value` to `new_state`, which is more informative?